File: _skels/next-js-skel/src/lib/seed.js
/**
 * Seed module — creates default user accounts from environment variables.
 *
 * Reads USER_* and SUPERUSER_* env vars and inserts the corresponding rows
 * into the `users` table if they do not already exist. Passwords are hashed
 * with bcrypt before storage. All DB access goes through Drizzle's query
 * builder -- no raw SQL.
 *
 * Called automatically by `getDb()` after schema initialisation so that
 * every first run of a fresh skeleton already has usable accounts.
 */

const bcrypt = require('bcryptjs');
const { eq } = require('drizzle-orm');
const { users } = require('./schema');

const SALT_ROUNDS = 12;

/**
 * Synchronously hash a password with bcrypt.
 * We intentionally use the sync variant here because better-sqlite3 is
 * synchronous and the seed runs once at startup -- there is no request
 * to block.
 */
function hashPasswordSync(password) {
  return bcrypt.hashSync(password, SALT_ROUNDS);
}

/**
 * Insert a user row if the username does not already exist.
 */
function upsertUser(db, { username, email, password }) {
  if (!username || !password) return;

  const existing = db
    .select({ id: users.id })
    .from(users)
    .where(eq(users.username, username))
    .get();
  if (existing) return;

  const hash = hashPasswordSync(password);
  db.insert(users)
    .values({
      username,
      email: email || null,
      password_hash: hash,
    })
    .run();
}

/**
 * Seed default accounts defined by environment variables.
 *
 * Reads directly from process.env rather than config to work reliably
 * in both standalone Node and Next.js server modes.
 */
function seedDefaultAccounts(db) {
  const accounts = [
    {
      username: process.env.USER_LOGIN || 'user',
      email: process.env.USER_EMAIL || 'user@example.com',
      password: process.env.USER_PASSWORD || 'secret',
    },
    {
      username: process.env.SUPERUSER_LOGIN || 'admin',
      email: process.env.SUPERUSER_EMAIL || 'admin@example.com',
      password: process.env.SUPERUSER_PASSWORD || 'secret',
    },
  ];
  for (const account of accounts) {
    upsertUser(db, account);
  }
}

module.exports = { seedDefaultAccounts };

File: _skels/rust-axum-skel/src/seed.rs
//! Seed default user accounts from environment variables at startup.
//!
//! Reads `USER_LOGIN`, `USER_EMAIL`, `USER_PASSWORD` and
//! `SUPERUSER_LOGIN`, `SUPERUSER_EMAIL`, `SUPERUSER_PASSWORD` from the
//! process environment and inserts each account into the `users` table
//! when one with that username does not already exist.

use chrono::Utc;
use sea_orm::{
    ActiveModelTrait, ColumnTrait, DatabaseConnection, EntityTrait, QueryFilter, Set,
};

use crate::auth::hash_password;
use crate::entities::user;
use crate::error::ApiError;

/// Account descriptor read from env vars.
struct SeedAccount {
    username: String,
    email: String,
    password: String,
}

/// Seed default accounts (regular user + superuser) into the database.
///
/// Each account is only created when the username does not already exist,
/// making the function safe to call on every startup.
pub async fn seed_default_accounts(db: &DatabaseConnection) -> Result<(), ApiError> {
    let accounts = collect_accounts();

    for acct in &accounts {
        if acct.username.is_empty() || acct.password.is_empty() {
            continue;
        }

        let existing = user::Entity::find()
            .filter(user::Column::Username.eq(&acct.username))
            .one(db)
            .await?;

        if existing.is_some() {
            tracing::info!(
                target: "rust_axum_skel",
                "[seed] User '{}' already exists — skipping",
                acct.username,
            );
            continue;
        }

        let hash = hash_password(&acct.password)?;
        let new_user = user::ActiveModel {
            username: Set(acct.username.clone()),
            email: Set(acct.email.clone()),
            password_hash: Set(hash),
            created_at: Set(Utc::now()),
            ..Default::default()
        };
        new_user.insert(db).await?;

        tracing::info!(
            target: "rust_axum_skel",
            "[seed] Created default user '{}'",
            acct.username,
        );
    }

    Ok(())
}

/// Collect seed accounts from environment variables.
fn collect_accounts() -> Vec<SeedAccount> {
    vec![
        SeedAccount {
            username: std::env::var("USER_LOGIN").unwrap_or_default(),
            email: std::env::var("USER_EMAIL").unwrap_or_default(),
            password: std::env::var("USER_PASSWORD").unwrap_or_default(),
        },
        SeedAccount {
            username: std::env::var("SUPERUSER_LOGIN").unwrap_or_default(),
            email: std::env::var("SUPERUSER_EMAIL").unwrap_or_default(),
            password: std::env::var("SUPERUSER_PASSWORD").unwrap_or_default(),
        },
    ]
}
